File: Dart/Add Two Numbers.dart
/**
 * Definition for singly-linked list.
 */
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    return help(l1, l2, 0);
  }

  ListNode? help(ListNode? l1, ListNode? l2, int carry) {
    if (l1 == null && l2 == null) {
      return carry == 0 ? null : ListNode(carry);
    }
    if (l1 == null) {
      return help(l2, ListNode(carry), 0);
    }
    if (l2 == null) {
      return help(l1, ListNode(carry), 0);
    }
    int sum = l1.val + l2.val + carry;
    print(sum % 10);
    ListNode? next = help(l1.next, l2.next, sum ~/ 10);
    return ListNode(sum % 10, next);
  }

  void printList(ListNode? node) {
    while (node != null) {
      print(node.val);
      node = node.next;
    }
  }
}

void main() {
  Solution solution = Solution();

  // Create first number: 342 (represented as 2 -> 4 -> 3)
  ListNode l1 = ListNode(2, ListNode(4, ListNode(3)));

  // Create second number: 465 (represented as 5 -> 6 -> 4)
  ListNode l2 = ListNode(5, ListNode(6, ListNode(4)));

  // Add the two numbers
  ListNode? result = solution.addTwoNumbers(l1, l2);
  print("result: \n");
  // Print the result: 807 (represented as 7 -> 0 -> 8)
  solution.printList(result);
}

File: Python/add-two-numbers.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        result: Optional[ListNode] = ListNode(-1)
        l1_runner = l1
        l2_runner = l2
        carry: int = 0
        pred = result
        while l1_runner or l2_runner or carry:
            if l1_runner and l2_runner:
                sum = l1_runner.val + l2_runner.val + carry
                l1_runner.val = sum % 10
                carry = sum // 10
                pred.next = l1_runner
                pred = pred.next
                l1_runner = l1_runner.next
                l2_runner = l2_runner.next
                continue
            if l1_runner:
                sum = l1_runner.val + carry
                l1_runner.val = sum % 10
                carry = sum // 10
                pred.next = l1_runner
                pred = pred.next
                l1_runner = l1_runner.next
                continue
            if l2_runner:
                sum = l2_runner.val + carry
                l2_runner.val = sum % 10
                carry = sum // 10
                pred.next = l2_runner
                pred = pred.next
                l2_runner = l2_runner.next
                continue
            if carry:
                pred.next = ListNode(1)
                break

        return result.next


def list_to_linkedlist(lst):
    dummy = ListNode()
    curr = dummy
    for val in lst:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next


def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result


if __name__ == "__main__":
    l1 = list_to_linkedlist([2, 4, 3])
    l2 = list_to_linkedlist([5, 6, 4])
    sol = Solution()
    result = sol.addTwoNumbers(l1, l2)
    print(linkedlist_to_list(result))
